Source/DafnyCore/Compilers/ResolvedDesugaredExecutableDafny/Dafny-compiler-fdafny.dfy

```diff
@@ -2,15 +2,209 @@ module {:extern "ResolvedDesugaredExecutableDafnyPlugin"} ResolvedDesugaredExecu
+    // Methods are just functions
+    static method EmitClass(c: Class) returns (s: DAM.Stmt) decreases * {
+      var fields := map[];
+      // TODO: normal fields turn into references
+      /*for i := 0 to |c.fields| {
+        match c.fields[i]
+        case Field(name, defaultValue) =>
+          fields := fields[name := DAM.Expr.Ref(new DAM.Ptr())];
+        case _ => UnsupportedFeature.Throw(); s := DAM.Skip();
+      }*/
+      for i := 0 to |c.body| {
+        match c.body[i]
+        case Method(m) =>
+          var meth := EmitMethod(m);
+          fields := fields[m.name := meth];
+      }
+      s := DAM.Stmt.Record(fields);
+    }
+
+    static method EmitMethod(m: Method) returns (s: DAM.Stmt) decreases * {
+      // Label a point for early returns
+      var body := EmitBlock(m.body);
+      s := DAM.LetCS("return", body);
+      // Out parameters turn into input references
+      /*for i := 0 to |m.outVars| {
+        match m.outVars
+        case Ident(Ident(v)) =>
+        
+      }*/
+      // In parameters...also turn into input references
+      for i := 0 to |m.params| {
+        match m.params[|m.params| - i - 1]
+        case Formal(arg, dom) =>
+          var dom := PolarizePos(dom);
+          s := DAM.Func(arg, dom, s);
+      }
+      // Annotate the entire function to allow it to appear in a redex
+    }
+
+    // Dafny expressions are actually DAM statements, since they are sensitive to evaluation order
+    static method EmitExpr(e: Expression) returns (s: DAM.Stmt) {
+      match e
+      case Ident(v) => s := DAM.Pure(DAM.Var(v));
+      case This()   => s := DAM.Pure(DAM.Var("this"));
+      case Ite(cond, thn, els) =>
+        var cond := EmitExpr(cond);
+        var thn  := EmitExpr(thn);
+        var els  := EmitExpr(els);
+        s := DAM.Stmt.Bind(cond, "if", DAM.Ite(DAM.Var("if"), thn, els));
+      case _ => UnsupportedFeature.Throw(); s := DAM.Skip();
+    }
+
+    // Terminates by multiset ordering on block
+    static method EmitBlock(block: seq<Statement>) returns (st: DAM.Stmt) decreases * {
+      if (|block| <= 0) { return DAM.Skip(); }
+      var next := block[1..];
+      match block[0]
+
+      case If(cond, thn, els) =>
+        var cond := EmitExpr(cond);
+        var thn  := EmitBlock(thn);
+        var els  := EmitBlock(els);
+        var next := EmitBlock(next);
+        st := DAM.Then(DAM.Stmt.Bind(cond, "if", DAM.Ite(DAM.Var("if"), thn, els)), next);
+      
+      case While(guard, body) =>
+        var guard := EmitExpr(guard);
+        var body := EmitBlock(body);
+        var next := EmitBlock(next);
+        st := DAM.While(guard, body, next);
+      
+      // Breaks, returns, and halts drop the rest of the block
+      case Break(lab) => {
+        match lab
+        case Some(lab) =>
+          st := DAM.Throw(DAM.Var(lab), DAM.Skip());
+        case None => UnsupportedFeature.Throw(); st := DAM.Skip();
+      }
```

Comment:
Instead of putting `{ ... }` around the entire `match lab ...` statement, put the `{ ... }` around the list of `case`s, like this:

``` dafny
match lab {
  case Some(lab) =>
    ...
  case None => ...
}
```